File: test/nodejs/test_harness.js
// Based on:
//    https://code.djangoproject.com/browser/django/trunk/tests/regressiontests/templates/tests.py
//
var sys = require('util'), fs = require('fs');

var Thistle = require('../../nodejs/thistle');

// Test classes
function SomeException() {
    Error.apply(this, arguments);
}
SomeException.prototype = new Error();
SomeException.prototype.constructor = SomeException;
SomeException.prototype.name = "SomeException";

function SomeOtherException() {
    Error.apply(this, arguments);
}
SomeOtherException.prototype = new Error();
SomeOtherException.prototype.constructor = SomeOtherException;
SomeOtherException.prototype.name = "SomeOtherException";

//
//
//
function SomeClass() {
    this.otherclass = new OtherClass();
};

SomeClass.prototype = {
    method: function () {
        return "SomeClass.method";
    },

    method2 : function(o) {
        return o;
    },

    method3 : function() {
        throw new SomeException();
    },

    method4 : function() {
        throw new SomeOtherException();
    }
};

function SilentGetItemClass() {
};

SilentGetItemClass.prototype = {
};

function SilentAttrClass() {
};

SilentAttrClass.prototype = {
};

function OtherClass() {
}

OtherClass.prototype = {
    method : function() {
        return "OtherClass.method";
    }
};

function TestObj() {
};

TestObj.prototype = {
    is_true : function() { return true; },
    is_false : function() { return false; },
    is_bad : function() { don_run_me(); },
};

//
//
//
function removeComments(text) {
    var idx;
    while ((idx = text.indexOf("/*")) != -1) {
        var eidx = text.indexOf("*/", idx);
        eidx = (eidx == -1) ? text.length : eidx + 2;

        text = text.slice(0, idx) + text.slice(eidx);
    }
    while ((idx = text.indexOf("//")) != -1) {
        var eidx = text.indexOf("\n", idx);
        eidx = (eidx == -1) ? text.length : eidx + 1;

        text = text.slice(0, idx) + text.slice(eidx);
    }
    return text;
}

//
exports.testBasic = function(test) {
    var data = fs.readFileSync("input/basic_template.json");

    var tests = JSON.parse(removeComments(""+data));

    Thistle.template_loaders = [
        function(name, dirs) {
            // return [tests[name][0], name];
            return [new Thistle.Template(tests[name][0], null, name), name];
        },
    ];

    for (var tcase in tests) {
        // if (tcase != 'include10') continue;
        // if (tcase != 'autoescape-tag08') continue;
        // if (tcase != 'if-tag-eq01') continue;

        var tdata = tests[tcase];
        var expect = tdata[2];

        Thistle.TEMPLATE_STRING_IF_INVALID = "INVALID";
        sys.puts("Running: " + tcase);

        try {
            var func = function() {
                var tmpl = new Thistle.Template(tdata[0]);
                return tmpl.render(new Thistle.Context(tdata[1]));
            };

            //sys.puts(func());

            if (expect == "Thistle.ParseException") {
                test.throws(func, Thistle.ParseException, tcase);
            } else if (expect == "Thistle.TemplateSyntaxError") {
                test.throws(func, Thistle.TemplateSyntaxError, tcase);
            } else if (expect == "Thistle.TemplateDoesNotExist") {
                test.throws(func, Thistle.TemplateDoesNotExist, tcase);
            } else {
                var val = func();

                if (!(tdata[2] instanceof Array)) {
                    test.equal(tdata[2], val, tcase);
                } else {
                    var found = false;
                    for (var i = 0; i < tdata[2].length; i++) {
                        if (tdata[2][i] == val) {
                            test.equal(tdata[2][i], val, tcase);
                            found = true;
                        }
                    }
                    if (!found)
                        test.equal(tdata[2], val, tcase);
                }
            }
        } catch (e) {
            if (expect == "Thistle.ParseException") {
                test.throws(func, Thistle.ParseException, tcase);
            } else if (expect == "Thistle.TemplateSyntaxError") {
                test.throws(func, Thistle.TemplateSyntaxError, tcase);
            } else if (expect == "Thistle.TemplateDoesNotExist") {
                test.throws(func, Thistle.TemplateDoesNotExist, tcase);
            } else if (expect instanceof Array) {
                for (var i = 0; i < expect.length; i++) {
                    if (expect[i] == "Thistle.TemplateDoesNotExist") {
                        // TODO - 
                    }
                }
            } else {
                sys.puts('EXCEPTION = ' + e);
                if (e != undefined) {
                    sys.puts("TYPE = " + typeof e + "Is TemplateSyntax = " + e instanceof Thistle.TemplateSyntaxError);
                    sys.puts(e.name);
                    sys.puts("MESSAGE: " + e.message);
                }
                sys.puts("EXCEPTION in " + tcase);
                throw e;
            }
        }
    }

    test.done();
};

/*
exports.testFile = function(test) {
    var tmpl_data = fs.readFileSync('../input/test1.tmpl');
    var json_data = JSON.parse(fs.readFileSync('../input/test1.json'));
    var expected  = fs.readFileSync('../output/test1.html');

    var thistle = new Thistle(tmpl_data.toString());
    test.equal(thistle.render(json_data), expected.toString(), "First test");
    test.done();
};
*/

File: test/py/test_harness.py
# -*- coding: utf-8 -*-

import thistle

import unittest
from datetime import datetime, timedelta
import time
import os
import sys
import traceback
import warnings
from urlparse import urljoin

#####################################
# Helper objects for template tests #
#####################################

class SomeException(Exception):
    silent_variable_failure = True

class SomeOtherException(Exception):
    pass

class ContextStackException(Exception):
    pass

class SomeClass:
    def __init__(self):
        self.otherclass = OtherClass()

    def method(self):
        return "SomeClass.method"

    def method2(self, o):
        return o

    def method3(self):
        raise SomeException

    def method4(self):
        raise SomeOtherException

    def __getitem__(self, key):
        if key == 'silent_fail_key':
            raise SomeException
        elif key == 'noisy_fail_key':
            raise SomeOtherException
        raise KeyError

    def silent_fail_attribute(self):
        raise SomeException
    silent_fail_attribute = property(silent_fail_attribute)

    def noisy_fail_attribute(self):
        raise SomeOtherException
    noisy_fail_attribute = property(noisy_fail_attribute)

class OtherClass:
    def method(self):
        return "OtherClass.method"

class TestObj(object):
    def is_true(self):
        return True

    def is_false(self):
        return False

    def is_bad(self):
        time.sleep(0.3)
        return True

class SilentGetItemClass(object):
    def __getitem__(self, key):
        raise SomeException

class SilentAttrClass(object):
    def b(self):
        raise SomeException
    b = property(b)

class UTF8Class:
    "Class whose __str__ returns non-ASCII data"
    def __str__(self):
        return u'ŠĐĆŽćžšđ'.encode('utf-8')

class Templates(unittest.TestCase):
    def test_templates(self):
        template_tests = self.get_template_tests()

        thistle.Thistle.TEMPLATE_STRING_IF_INVALID = "INVALID"

        def loader(name, dirs):
            return (thistle.Template(template_tests[name][0], None, name), name)

        thistle.Template.template_loaders.append(loader)

        failures = []

        for name, vals in template_tests.items():
            #if name != 'basic-syntax27': continue

            if not vals : continue
            print("Running: %s" % name)

            
            try:
                found = False

                def func():
                    tmpl = thistle.Template(vals[0])
                    return tmpl.render(thistle.Context(vals[1]))

                result = vals[2]
                output = func()
                if output == result:
                    found = True
                elif isinstance(result, (tuple, list)):
                    if output in vals[2]:
                        found = True

                if not found:
                    failures.append("Template test: %s -- FAILED. Expected %r, got %r" % (name, result, output))
                    print failures[-1]
            except Exception as e:
                if type(vals[2]) == type(object) or vals[2] == "Thistle.TemplateSyntaxError":
                    #if issubclass(thistle.ThistleException, e.__class_):
                    #    pass
                    pass
                else:
                    failures.append("Template test: %s -- FAILED. Expected %r, got %r" % (name, result, e))
                    print failures[-1]
                pass

    def get_template_tests(self):
        import  json, re

        def removeComments(text):
            while text:
                idx = text.find('/*')
                if idx == -1:
                    break
                eidx = text.find('*/', idx)
                if eidx == -1:
                    eidx = len(text)
                text = text[0:idx] + text[eidx+2:]
            while text:
                idx = text.find('//')
                if idx == -1:
                    break
                eidx = text.find("\n", idx)
                if eidx == -1:
                    eidx = len(text)
                text = text[0:idx] + text[eidx+1:]
            #if text.find("END_OF_FILE"):
            #    text = text[0:text.find("END_OF_FILE")] + '"eof":[]}'
            return text

        with open('input/basic_template.json') as fd:
           data = removeComments(fd.read())
           return  json.loads(data)

if __name__ == "__main__":
    unittest.main()
